File: 147InsertionSortList.cpp
//1. Keep top/prev/cur pointer for a fixed place. h pointer to loop over top ->pre to compare with h.
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode* cur = head;
        ListNode a(-1);
        ListNode* h = &a;
        h->next = head;
        ListNode* top = h, *prev = h;
        int n = 0;
        int needStep = 0;
        while(cur){
            if(n==0) {cur = cur->next; prev = prev->next; needStep++; n=needStep; h = top;} //If we can not find.
            else{
                if(cur -> val >= h ->next -> val) {n--;h = h->next;}
                else{
                    prev -> next = cur -> next;
                    cur-> next = h -> next;
                    h -> next = cur;
                    cur = prev -> next;
                    needStep++; n=needStep; h = top;
                }
            }
        }
        return top->next;
    }
};


//Consice Version.
class Solution {
public:
    ListNode* insertionSortList(ListNode* head) {
        ListNode* dummy = new ListNode(0);
        dummy -> next = head;
        ListNode *pre = dummy, *cur = head;
        while (cur) {
            if ((cur -> next) && (cur -> next -> val < cur -> val)) {
                while (pre -> next -> val < cur -> next -> val) {
                    pre = pre -> next;
                }
                ListNode* temp = pre -> next;
                pre -> next = cur -> next;
                cur -> next = cur -> next -> next;
                pre -> next -> next = temp;
              
                pre = dummy;
                //cur = cur->next;
            }
            else {
                cur = cur -> next;
            }
        }
        return dummy -> next;
    }
};

File: Algo/Leetcode/147InsertionSortList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def insertionSortList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        f = ListNode()

        while head is not None:
            tmp = head.next
            head.next = None
            self.insert(f, head)
            head = tmp

        return f.next

    def insert(self, f, head):
        m = f
        while m.next and m.next.val < head.val:
            m = m.next

        head.next = m.next
        m.next = head
